shared/extension/NewScript.hpp

```diff
@@ -0,0 +1,285 @@
+            Target& operator=(const Target& other) = default;
+            Target& operator                       =(Target&& other) noexcept {
+                id       = other.id;
+                position = other.position;
+                gain     = other.gain;
+                torque   = other.torque;
+                return *this;
+            }
+
+            /// @brief The servo ID that this target is for; see ServoID.hpp
+            ServoID id;
+            /// @brief Target position for this servo in this frame
+            float position{0.0f};
+            /// @brief Gain for this servo in this frame
+            float gain{0.0f};
+            /// @brief Torque for this servo in this frame
+            float torque{0.0f};
+        };
+
+        Frame() : duration() {}
+        Frame(const NUClear::clock::duration& dur, std::vector<Target> targets)
+            : duration(dur), targets(std::move(targets)) {}
+
+        /// @brief The time it should take for the servos to move to their targets
+        NUClear::clock::duration duration;
+        /// @brief Target position of servos for this frame
+        std::vector<Target> targets;
+    };
+
+    template <typename Sequence>
+    struct Script {
+        std::vector<Frame> frames;
+
+        /// @brief Assumes the hostname is of the form <platform_name><robot_number> and returns <platform_name>.
+        /// @param hostname The hostname of the device the program is running on.
+        /// @return The platform name.
+        /// @throws System error when the platform cannot be extracted from the hostname.
+        static inline std::string getPlatform(const std::string& hostname) {
+            // It is assumed that all hostnames are in the format <platform name><robot number>,
+            // such that the regular expression
+            // [a-z]+[0-9]+?
+            // will match all hostnames
+            std::regex re("([a-z]+)([0-9]+)?");
+            std::smatch match;
+
+            if (std::regex_match(hostname, match, re)) {
+                // match[0] will be the full string
+                // match[1] the first match (platform name)
+                // match[2] the second match (robot number)
+                return match[1].str();
+            }
+
+            throw std::system_error(-1,
+                                    std::system_category(),
+                                    ("Failed to extract platform name from '" + hostname + "'."));
+        }
+
+        /// @brief If a robot-specific script exists, load the YAML file as a script. If not, loads the
+        /// platform-specific script. If neither exist, throws a runtime error.
+        /// @param script The name of the script to run.
+        /// @return The script contents in the form off a YAML node
+        /// @throws Runtime error when the script doesn't exist.
+        static YAML::Node load_script(const std::string& script) {
```

Comment:
```suggestion
        static YAML::Node load(const std::string& script) {
```
Doubling up on the name script, `Script::load_script` vs `Script::load`.